gematria/granite/python/token_graph_builder_model.py

```diff
@@ -340,3 +398,72 @@ def _create_graph_network_modules(
             residual_connection=options.EnableFeature.BY_FLAG,
         ),
     )
+
+  def _make_batch_feed_dict(self) -> model_base.FeedDict:
+    feed_dict = super()._make_batch_feed_dict()
+
+    feed_dict[self._instruction_annotations] = (
+        self._batch_graph_builder.instruction_annotations
+    )
+    return feed_dict
+
+
+class TokenGraphBuilderModelNodeEmbed(snt.Embed):
+  """Extends `snt.Embed` to include instruction annotations in node embeddings.
+
+  Generates node embeddings normally, then replaces the last `num_annotation`
+  values of the embeddings corresponding to instructions with the annotation
+  values. The embeddings for other node types remain unchanged.
+  """
+
+  def __init__(
+      self,
+      instruction_annotations,
+      instruction_node_mask,
+      **kwargs,
+  ) -> None:
+    """Initializes node embeddings.
+
+    Args:
+      instruction_annotations: Tensor holding instruction level runtime
+        annotations as in `BasicBlockGraphBuilder`.
+      instruction_node_mask: As in `BasicBlockGraphBuilder`.
+      **kwargs: Additional arguments to be passed to the internal `snt.Embed`.
+    """
+    super().__init__(**kwargs)
+
+    # The number of annotations per instruction.
+    self.num_annotations = int(instruction_annotations.shape[1])
+
+    if self.num_annotations > self.embed_dim:
+      raise ValueError('num_annotations cannot be greater than embed_dim.')
+
+    self.instruction_annotations = instruction_annotations
+    self.instruction_node_mask = instruction_node_mask
+
+  def __call__(self, inputs):
+    embeddings = super().__call__(inputs)
+
+    if self.num_annotations == 0:
+      return embeddings
+
+    out = tf.concat(
+        [
+            tf.slice(
+                embeddings,
+                begin=[0, 0],
+                size=[-1, self.embed_dim - self.num_annotations],
```

Comment:
Ideally, we should put this to the model (i.e. either compute the learned embedding size from a specified node feature size + annotation size, or compute the node state size from a specified learned embedding size + annotation size).

Reply:
Sorry, I'm not sure I understand what is to be changed - this snippet should update the last `num_annotations` with the annotations for instruction nodes and keep the rest `embed_dim - num_annotations` intact, with `embed_dim` and `num_annotations` both coming from the model.

Reply:
Ah, got it.

Regarding the slicing - the reason why it currently isn't as clean as a simple concatenate operation is that we are only updating instruction nodes, so we'd have to handle the last `num_annotation` entries of non-instruction nodes separately.

A possible solution would be to have two `snt.Embed`s composed in this class - one holding the learnt embeddings of length `embed_dim - num_annotations` for all nodes, and another holding learnt embeddings of length `num_annotations` for only non-instruction nodes, which would be interleaved with `instruction_annotations` accordingly to the `instruction_node_mask` and then concatenated with the other one.

Or instead of having the second set of learnt embeddings we could just use a larger `embed_dim` and let the last `num_annotation` embedding entries for non-instruction nodes be something like zeros - though that sounds like it might be wasteful.